@here/harp-mapview/lib/PolarTileDataSource.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as THREE from "three";
+
+import { StyleSet } from "@here/harp-datasource-protocol";
+import { MapEnv, StyleSetEvaluator } from "@here/harp-datasource-protocol/index-decoder";
+import {
+    GeoCoordinates,
+    MathUtils,
+    MercatorProjection,
+    polarTilingScheme,
+    TileKey,
+    TilingScheme
+} from "@here/harp-geoutils";
+
+import { DataSource } from "./DataSource";
+import { createMaterial } from "./DecodedTileHelpers";
+import { Tile } from "./Tile";
+
+/**
+ * [[DataSource]] providing geometry for poles
+ */
+export class PolarTileDataSource extends DataSource {
```

Comment:
I like the idea of encapsulating the pole caps into a dedicated datasource but I think we cannot do it because:
1. We will have a lot of draw calls b/c there will be a lot of tiny tiles at the pole (compared to one draw call if you generate dedicated geometry)
2. We force the VisibleTileSet to traverse the tile tree twice, once for webMercatorTilingScheme and once for the polarTilingScheme (btw. we also call it HereTilingScheme b/c this is what we used in the past for our proprietary map formats).
3. The geometry will not match the webMercator tiles

Reply:
Thanks for reivew. Yes, there are issues to this solution, it's just the best one I've found so far.
 1. This is kind of a tradeoff because from one side there are cpu and memory complexity with dedicated geometry (tons of subdivisions at higher levels - 2^20 actually), and gpu complexity from the other with tons of quads. But I think this solution has a bit more scalability.
 2. Yep, this is a bad one. But the amount of tiles can be reduced significantly if we could disable subdivision by latitude. I have tried to develop such subdivision scheme, but the SubTiles iterator seems to be not able to create tiles with no subdivision (e.g. with getSubdivisionY() => 1) properly.
 3. Yes, it will not match, and there are no tiling scheme that can both address latitudes >85 and math mercator tiles. But as long as webMercator tiling has same horizontal subdivision scheme, they will match at least horizontally.

Reply:
Just checked, zooming in directly to poles makes things really bad.

Reply:
If we encounter that the triangles for a dedicated geometry are to many we could do some smart culling that renders only sub-ranges of that geometry (i.e. a segment of the circle instead of the whole circle)
I think we should limit the tile level close to the poles somehow. Already rendering the "real" data at level 20 close to the pole will kill us with to many draw calls
Since the raw data is at level 14 the deeper levels are just clipped duplicates of the same data we could make the max storage level that is used depending on the latitude.

Reply:
Note how google maps does not give a hang about this problem :sob: https://www.google.fr/maps/@-85,149.3853301,5.69z

Reply:
@Astrak interesting. And when you zoom in closer than 10 you cannot see the pole anymore. They force the camera to stay inside the valid webmercator range.